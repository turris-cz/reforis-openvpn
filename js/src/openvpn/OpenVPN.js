/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import { useAPIGet, Spinner, AlertContextProvider } from "foris";

import API_URLs from "API";
import AuthorityStatus from "./AuthorityStatus";
import ServerSettings from "./ServerSettings";

OpenVPN.propTypes = {
    ws: PropTypes.object.isRequired,
};

export default function OpenVPN({ ws }) {
    const [authorityResponse, getAuthority] = useAPIGet(API_URLs.authority);
    useEffect(() => {
        getAuthority();
    }, [getAuthority]);

    const [settingsResponse, getSettings] = useAPIGet(API_URLs.serverSettings);
    useEffect(() => {
        getSettings();
    }, [getSettings]);

    let componentContent;
    const authorityData = authorityResponse.data;
    const settingsData = settingsResponse.data;
    if (
        authorityResponse.isLoading
        || settingsResponse.isLoading
        || !authorityData
        || !settingsData
    ) {
        componentContent = <Spinner className="my-3 text-center" />;
    } else if (authorityResponse.isError || settingsResponse.isError) {
        componentContent = (
            <>
                {authorityResponse.isError && <p className="text-center text-danger">{_("An error occurred during loading certificate authority details")}</p>}
                {settingsResponse.isError && <p className="text-center text-danger">{_("An error occurred during loading OpenVPN settings")}</p>}
            </>
        );
    } else {
        componentContent = (
            <>
                <AuthorityStatus
                    ws={ws}
                    status={authorityData.status}
                    serverEnabled={settingsData.enabled}
                    onReload={getAuthority}
                >
                    <ServerSettings settingsData={settingsData} />
                </AuthorityStatus>
            </>
        );
    }

    return (
        <>
            <h1>{_("OpenVPN")}</h1>
            <AlertContextProvider>
                {componentContent}
            </AlertContextProvider>
        </>
    );
}
