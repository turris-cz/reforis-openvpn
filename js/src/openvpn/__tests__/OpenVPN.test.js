/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, waitForElement, fireEvent } from "customTestRender";
import mockAxios from 'jest-mock-axios';

import { WebSockets } from "foris";

import OpenVPN from "../OpenVPN";

describe("<OpenVPN />", () => {
    it("should render child and alert", async () => {
        const webSockets = new WebSockets();
        const { getByText, container } = render(<OpenVPN ws={webSockets} />);

        expect(getByText("OpenVPN")).toBeDefined();

        mockAxios.mockResponse({data: {status: "ready"}});
        await waitForElement(() => getByText("Certificate authority"));

        fireEvent.click(getByText("Delete CA"));
        mockAxios.mockError({response: {}});
        await waitForElement(() => getByText("Cannot delete certificate authority"));
        expect(container).toMatchSnapshot();
    });
});
