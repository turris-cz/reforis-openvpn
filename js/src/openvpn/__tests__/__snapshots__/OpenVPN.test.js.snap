// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OpenVPN /> should render child and alert 1`] = `
<div>
  <h1>
    OpenVPN
  </h1>
  <div
    class="alert alert-dismissible alert-danger"
  >
    <button
      class="close"
      type="button"
    >
      Ã—
    </button>
    Cannot delete certificate authority
  </div>
  <h3
    class="mb-3"
  >
    Server settings
  </h3>
  <p>
    Please note that you need a public (preferably static) IP address and your network configured to make use of VPN server.
  </p>
  <form
    class="mb-3"
  >
    <div
      class="col-sm-12 offset-lg-1 col-lg-10"
      style="margin-bottom: 1rem;"
    >
      <div
        class="custom-control custom-checkbox"
        style="margin-bottom: 0px;"
      >
        <input
          class="custom-control-input"
          id="1"
          type="checkbox"
        />
        <label
          class="custom-control-label"
          for="1"
        >
          Server enabled
        </label>
      </div>
    </div>
    <button
      class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
      type="submit"
    >
       
       
      Save
    </button>
  </form>
  <p>
    <strong>
      Advanced users:
    </strong>
     if you already configured OpenVPN server manually your config will be extended (rather than  overwritten). In case of conflict you must fix previous settings by yourself.
  </p>
  <h3>
    Certificate authority
  </h3>
  <p>
    Your certificate authority (CA) is set up properly.
  </p>
  <button
    class="btn btn-primary "
    type="button"
  >
     
     
    Delete CA
  </button>
</div>
`;
